src/lib.rs

```diff
@@ -67,29 +67,80 @@ impl AsciiChars {
     #[inline]
     pub fn find(self, haystack: &str) -> Option<usize> {
         let haystack = haystack.as_bytes();
-
-        let ptr = haystack.as_ptr();
-        let mut offset = 0;
         let mut len = haystack.len();
 
-        let mut res: usize;
+        if len == 0 { return None }
 
-        // Zero-length strings have a pointer set to 0x1. Even though the
-        // length is zero, we still trigger a bad access exception. I
-        // think this indicates that the instruction reads in 16 bytes
-        // worth of memory at a time, regardless of the length instruction.
-        //
-        // This could also be an indication of a subtle bug, where we
-        // might trigger access violations if we are near the end of a
-        // page. See the comment by Renat Saifutdinov on
-        // http://www.strchr.com/strcmp_and_strlen_using_sse_4.2
-        // It is suggested to use an "aligned read with mask false bits"
-        // to avoid the problem.
+        // The PCMPxSTRx instructions always read 16 bytes worth of
+        // data. To avoid walking off the end of a page (and
+        // potentially into a protected area), we read in 16-byte
+        // chunks aligned to the *end* of the string. The instructions
+        // handle unaligned access just fine. The trick lies in
+        // searching the leftover bytes at the beginning of the
+        // string.
         //
-        // We don't do this yet.
-        if len == 0 { return None }
+        // To do that, we use the PCMPESTRM instruction on the
+        // 16-byte-aligned block that contains the *start* of the
+        // string. This returns a mask of all the matching
+        // characters. We can can then ignore unrelated leading bits
+        // to find the index of the first related character (if any).
+
+        let true_ptr = haystack.as_ptr();
+
+        // Start at the 16-byte-aligned block *before* the string data
+        // starts
+        let ptr = (true_ptr as usize & !0xF) as *const u8;
 
-        loop {
+        // Find where the string really starts
+        let initial_offset = true_ptr as usize & 0xF;
+        let mut offset = initial_offset;
+
+        // If the string is magically aligned, skip this extra work
+        if offset != 0 {
```

Comment:
Yeah, I had debated that myself. Two half-suggestions seems like one full suggestion to me! I've pushed a new version.